### PR TITLE
Improve safety when using and combing app and protocol tree

### PR DESCRIPTION
**Pull request by anthony-murphy**
## Description
There is a bug in the code when a detached container with blobs is uploaded with the single commit summary flow. The bug results from combineAppAndProtocolSummary being called twice on the same summary. This change fixes that issue by improving typing and defensiveness when using and combing app and protocol trees, such that double wrapping can be detected, and thrown as an error. The fix for the specific issue is in `packages\loader\container-loader\src\protocolTreeDocumentStorageService.ts`. Additionally, i've updated a number of usage of the combined tree to remove unsafe casting.

Validated the issue via test modifications, which now pass with fixes.
Fixes [AB#3711](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/3711)

### CODE REVIEW

File: packages/drivers/odsp-driver/src/odspDocumentServiceFactoryCore.ts [hunk 1]
```diff
@@ -106,10 +107,9 @@ export class OdspDocumentServiceFactoryCore implements IDocumentServiceFactory {
 			throw new Error("A new or existing file must be specified to create container!");
 		}
 
-		const protocolSummary = createNewSummary?.tree[".protocol"];
-		if (protocolSummary) {
+		if (isCombinedAppAndProtocolSummary(createNewSummary)) {
 			const documentAttributes = getDocAttributesFromProtocolSummary(
-				protocolSummary as ISummaryTree,
+				createNewSummary.tree[".protocol"],
```
Comment (dannimad) [inline packages/drivers/odsp-driver/src/odspDocumentServiceFactoryCore.ts:112]: Is it safe to always ask for combined app protocol summary? In this case only protocol summary is required. I wonder if there is a case where protocol is there but app is not

Comment (anthony-murphy) [reply packages/drivers/odsp-driver/src/odspDocumentServiceFactoryCore.ts:112]: this is in the create new path, so it will always be a combined summary

**Comment by ChumpChief** [comment]
I'm seeking to put a cleaner split between loader and driver, and in particular driver-utils is full of items that span the layers.  Is there a way we can structure this to avoid usage in both?

**Comment by anthony-murphy** [comment]
> I'm seeking to put a cleaner split between loader and driver, and in particular driver-utils is full of items that span the layers. Is there a way we can structure this to avoid usage in both?

not obviously, as this has to do with the contract for create new. 

> I'm seeking to put a cleaner split between loader and driver, and in particular driver-utils is full of items that span the layers. Is there a way we can structure this to avoid usage in both?

This is a case where the loader layer implements a driver interface to add an adapter layer over the real driver, so the code here gets shared between the adapter and the drivers themselves. not sure where else this would go without duplicating code. i wish i could just add a ts file somewhere that was shared, and didn't need to export anything, but as far as i can tell thats not a thing. i marked the methods internal, which should make any future refactoring you do easier.

File: packages/loader/driver-utils/src/summaryForCreateNew.ts [hunk 2]
```diff
@@ -11,16 +12,56 @@ import {
...
+): summary is CombinedAppAndProtocolSummary {
+	if (
+		summary?.tree === undefined ||
+		summary.tree?.[".app"]?.type !== SummaryType.Tree ||
+		summary.tree?.[".protocol"]?.type !== SummaryType.Tree
+	) {
+		return false;
+	}
+	const treeKeys = Object.keys(summary.tree);
+	if (treeKeys.length !== 2) {
```
Comment (jatgarg) [inline packages/loader/driver-utils/src/summaryForCreateNew.ts:43]: Can't we have more than 2 and still have app and protocol summary? Where does the .logTail(I think there were more like this) lives? Are they within the app summary?

Comment (anthony-murphy) [reply packages/loader/driver-utils/src/summaryForCreateNew.ts:43]: this doesn't affect snapshots (download) only summaries(upload), and `.logTail` only exists in snapshost. everything managed by the runtime in under the `.app` tree. The summaries we validate here are always created by `combineAppAndProtocolSummary` which is just a few lines down in this file, and has a fixed format.

**Review by jatgarg — APPROVED** [review]

**Pull request merged**